### PR TITLE
More pets

### PR DESCRIPTION
**Pull request by dvoet**
https://broadinstitute.atlassian.net/browse/GAWB-2828

changes
- I got fed up with all the different types of Workbench*Email that extend WorkbenchEmail... there are times where we just don't know which one to create so need to pick one which is stupid. I don't think the distinction adds any value so I consolidated them all to a WorkbenchEmail case class.
- The identifier of a pet is now the user's id and google project
- Renamed and moved all the WorkbenchUserSerivceAccount* objects in workbench-libs to model.google and removed the WorkbenchUser prefix because they really have nothing to do with users
- Pets now live as a sub-context of the user's context, not in a separate pets ou. This means we don't need a petServiceAccount attribute on users because it is built into the hierarchy.

see also https://github.com/broadinstitute/workbench-libs/pull/45

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Make sure Swagger is updated if API changes
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/DirectorySubjectNameSupport.scala [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package org.broadinstitute.dsde.workbench.sam.directory
 
 import org.broadinstitute.dsde.workbench.model._
+import org.broadinstitute.dsde.workbench.model.google.GoogleProject
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/DirectorySubjectNameSupport.scala:4]: this file contains the crux moving pets under their owners and making them project specific

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/JndiDirectoryDAO.scala [hunk 2]
```diff
@@ -135,55 +128,61 @@ class JndiDirectoryDAO(protected val directoryConfig: DirectoryConfig)(implicit
...
+      ctx.bind(subjectDn(user.id), identityContext)
+      user
     } catch {
       case _: NameAlreadyBoundException =>
-        throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Conflict, s"identity with id ${subject.value} already exists"))
+        throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Conflict, s"identity with id ${user.id} already exists"))
     }
   }
 
-  override def getPetServiceAccountForUser(userId: WorkbenchUserId): Future[Option[WorkbenchUserServiceAccountEmail]] = {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/JndiDirectoryDAO.scala:157]: pet SAs implicitly belong to a user now. Loading a pet SA implies a user so no need for this function. Also no need for special add/remove below

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionRoutes.scala [hunk 3]
```diff
@@ -35,32 +35,32 @@ trait GoogleExtensionRoutes extends ExtensionRoutes with UserInfoDirectives {
...
-              val resource = Resource(ResourceTypeName(resourceTypeName), ResourceId(resourceId))
-              val resourceAndPolicyName = ResourceAndPolicyName(resource, AccessPolicyName(accessPolicyName))
-              pathEndOrSingleSlash {
-                post {
-                  complete {
-                    import GoogleModelJsonSupport._
-                    googleExtensions.synchronizeGroupMembers(resourceAndPolicyName).map { syncReport =>
-                      StatusCodes.OK -> syncReport
-                    }
+        pathPrefix("resource") {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionRoutes.scala]: whitespace

File: src/main/resources/swagger/api-docs.yaml [hunk 4]
```diff
@@ -75,18 +75,24 @@ paths:
       tags:
         - Admin
 
-  /api/google/user/petServiceAccount:
+  /api/google/user/petServiceAccount/{project}:
```
Comment (helgridly) [inline src/main/resources/swagger/api-docs.yaml:78]: does this mean the old API is gone? if so, rolling out this version of sam will break leo...

Comment (dvoet) [reply src/main/resources/swagger/api-docs.yaml:78]: Old api is still there just undocumented. It doesn't have google in the uri.

**Comment by rtitle** [comment]
Wow nice! I'll review this and test it out with Leo.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala [hunk 5]
```diff
@@ -69,126 +69,113 @@ class GoogleExtensions(val directoryDAO: DirectoryDAO, val accessPolicyDAO: Acce
...
+  private def getPetServiceAccountsForUser(userId: WorkbenchUserId): Future[Seq[PetServiceAccount]] = {
+    directoryDAO.getAllPetServiceAccountsForUser(userId)
   }
 
-  private def setUpServiceAccount(user: WorkbenchUser, petServiceAccount: WorkbenchUserServiceAccount): Future[WorkbenchUserServiceAccountEmail] = {
+  private def setUpServiceAccount(petServiceAccount: PetServiceAccount): Future[PetServiceAccount] = {
     for {
-    // add Service Account User role to the configured emails so they can assume the identity of the pet service account
-      _ <- Future.traverse(petServiceAccountConfig.serviceAccountUsers) { email =>
-        googleIamDAO.addServiceAccountUserRoleForUser(petServiceAccountConfig.googleProject, petServiceAccount.email, email)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala:169]: @rtitle note this change - we are no longer adding ServiceAccountUser role. Instead we rely on the leo service account having project editor role at the org level.

Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala:169]: Yes; that makes sense.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/schema/JndiSchemaDAO.scala [hunk 6]
```diff
@@ -217,37 +204,33 @@ class JndiSchemaDAO(protected val directoryConfig: DirectoryConfig)(implicit exe
...
     val schema = ctx.getSchema("")
 
-    createAttributeDefinition(schema, "1.3.6.1.4.1.18060.0.4.3.2.400", Attr.petServiceAccount, "pet service account of the user", true)
+    createAttributeDefinition(schema, "1.3.6.1.4.1.18060.0.4.3.2.70", Attr.project, "google project", true, equality = Option("caseIgnoreMatch"))
 
     val attrs = new BasicAttributes(true) // Ignore case
-    attrs.put("NUMERICOID", "1.3.6.1.4.1.18060.0.4.3.2.300")
-    attrs.put("NAME", "workbenchPerson")
+    attrs.put("NUMERICOID", "1.3.6.1.4.1.18060.0.4.3.2.700")
+    attrs.put("NAME", "petServiceAccount")
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/schema/JndiSchemaDAO.scala]: Should there be an `Attr.petServiceAccount`? I think I saw it hardcoded in a few places.

**Comment by rtitle** [comment]
A lot of changes here but it all looks reasonable to me. I haven't tested it with Leo yet.

**Comment by rtitle** [comment]
It also occurs to me that we’ll need to make sure the switcheroo (pet -> user) still works, in Rawls and in Sam. We should have swat tests covering that.

**Comment by dvoet** [comment]
Yes, will take care of it on merge to develop

**Comment by jmthibault79** [comment]
@dvoet current Swat switcheroo tests for Sam and Rawls are here: https://github.com/broadinstitute/firecloud-ui/tree/develop/automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api

**Review by helgridly — COMMENTED** [review]
I can't responsibly thumb this because I still don't have my head wrapped around ldap-ness. But the route and service implementations seem right. In lieu of a thumb I will check it instead ✅

**Review by UrsaStutsman — APPROVED** [review]
LGTM

**Pull request merged**